src/DynamoCore/DSEngine/NodeToCode.cs

```diff
@@ -1,16 +1,692 @@
+                    MarkConnectivityForNode(selection, connectivityMatrix, nodeMap, path);
+                }
+            }
+            else
+            {
+                var nodes = path.ToList();
+                var idx1 = nodeMap[path.First()];
+
+                int k = nodes.FindIndex(n => !n.IsConvertible || !selection.Contains(n));
+                if (k >= 0)
+                {
+                    for (int i = k + 1; i < nodes.Count; i++)
+                    {
+                        int idx2;
+                        if (!nodeMap.TryGetValue(nodes[i], out idx2))
+                            continue;
+
+                        connectivityMatrix[idx1, idx2] = false;
+                        connectivityMatrix[idx2, idx1] = false;
+                    }
+                }
+            }
+
+            path.RemoveAt(path.Count - 1);
+        }
+
+        /// <summary>
+        /// For a selection, output partitions that each group of node can be
+        /// converted to code.
+        /// </summary>
+        /// <param name="selection"></param>
+        /// <returns></returns>
+        public static List<List<NodeModel>> GetCliques(
+            IEnumerable<NodeModel> selection)
+        {
+            var selectionSet = new HashSet<NodeModel>(selection);
+            var convertibleNodes = selection.Where(n => n.IsConvertible).ToList();
+            var count = convertibleNodes.Count;
+
+            var nodeDict = convertibleNodes.Zip(Enumerable.Range(0, count), (n, i) => new { n, i})
+                                           .ToDictionary(x => x.n, x => x.i);
+
+            // Connectivity matrix represents whether two nodes are allowed to
+            // be in the same code block or not. 
+            bool[,] connectivityMatrx = new bool[count, count];
+            for (int i = 0; i < count; i++)
+            {
+                for (int j = 0; j < count; j++)
+                {
+                    connectivityMatrx[i, j] = true;
+                }
+            }
+            
+            // For two convertible nodes if there is path from one to the other
+            // goes through some node that is unconvertible or not in the 
+            // selection set, these two nodes cannot be in the same coee block,
+            // otherwise there will be a circular reference. For example:
+            //
+            //     a -> x -> b
+            //
+            // And the final graph would be
+            //
+            //    a, b <--> x
+            // 
+            // So here for a node which is convertible and from selection set, 
+            // find all its upstream paths, and if any path contains node that 
+            // is not in the selection set or is not convertible, mark all 
+            // nodes after this node along path as not reachable. 
+            foreach (var node in convertibleNodes)
+            {
+                var path = new List<NodeModel>();
+                path.Add(node);
```

Comment:
ReSharper suggests that this be turned into `var path = new List<NodeModel>(node);` :)